### PR TITLE
Implement stepper for Puyo

### PR DESCRIPTION
**Pull request by frostburn**
Implement stepper classes for turn based Puyo and add them to the engine factory.

### CODE REVIEW

File: lib/common/engine.js [hunk 1]
```diff
@@ -10,10 +11,18 @@ const defaultOptions = {
   stateCacheSize: STATE_CACHE_SIZE,
 };
 
+function factory(name) {
```
Comment (RauliL) [inline lib/common/engine.js]: This should be in it's own file called `engine-factory.js`.

File: lib/common/engine.js [hunk 2]
```diff
@@ -10,10 +11,18 @@ const defaultOptions = {
...
+  let stepper = panelLeagueFactory(name);
```
Comment (RauliL) [inline lib/common/engine.js]: Blank line after local variable declaration.

File: lib/common/engine.js [hunk 3]
```diff
@@ -10,10 +11,18 @@ const defaultOptions = {
...
+  if (!stepper) {
+    stepper = puyoFactory(name);
+  }
+  return stepper;
```
Comment (RauliL) [inline lib/common/engine.js]: Blank line before `return` statement.

File: lib/common/puyo/basic.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+  for (let i = 0; i < blocks.length; ++i) {
+    const color = blocks[i];
+
+    if (color > emptyPuyo) {
+      const indices = extractGroup(blocks, i, width, height);
+
+      groups.push({ color, indices });
+    }
+  }
+  return groups;
```
Comment (RauliL) [inline lib/common/puyo/basic.js]: Blank line before `return` statement.

File: lib/common/engine.js (hunk 1 above)
Comment (frostburn) [inline lib/common/engine.js]: A whole file for a single function whose only purpose is to import a few other files?

**Comment by frostburn** [comment]
Do we have a style guide somewhere? I don't find the suggested whitespace pretty and the linter won't complain.

File: lib/common/engine.js (hunk 1 above)
Comment (RauliL) [inline lib/common/engine.js]: I find it neater design.

**Comment by frostburn** [comment]
Issues agreed to be minor. The whole codebase needs a style & lint overhaul. Deferring this for later.

**Pull request merged**